### PR TITLE
PERF: Split blocks in blk.delete

### LINKED ISSUE #46267: PERF: performance regression in assigning value to df[col_name]
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this issue exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this issue exists on the main branch of pandas.

### Reproducible Example

```python
import numpy as np
import pandas as pd
from tqdm import tqdm

df = pd.DataFrame(np.random.random((5000,15000)))
df.columns = df.columns.astype("str")
col_0_np = df.iloc[:, 0].to_numpy()

for idx in tqdm(df.columns):
    df[idx] = col_0_np
```

```
  0%|          | 20/15000 [00:04<57:50,  4.32it/s] 
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 06d230151e6f18fdb8139d09abf539867a8cd481
python           : 3.8.12.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19041
machine          : AMD64
processor        : AMD64 Family 23 Model 113 Stepping 0, AuthenticAMD
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : ***

pandas           : 1.4.1
numpy            : 1.19.2
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 21.2.2
setuptools       : 58.0.4
Cython           : None
pytest           : 6.2.5
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.8.0
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.0.2
IPython          : 7.31.1
pandas_datareader: None
bs4              : 4.8.0
bottleneck       : 1.3.2
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : 3.5.1
numba            : 0.54.1
numexpr          : 2.7.3
odfpy            : None
openpyxl         : 3.0.9
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.6.2
sqlalchemy       : 1.4.27
tables           : 3.6.1
tabulate         : None
xarray           : None
xlrd             : 2.0.1
xlwt             : None
zstandard        : None

</details>

### Prior Performance

In `1.3.5`, the code should complete within seconds.

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #46267 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

```
       before           after         ratio
     [a047fb6c]       [e8c3a52a]
-      1.16±0.01s         962±10μs     0.00  indexing.Setitem.time_setitem_list
-         382±7ms         302±10μs     0.00  indexing.Setitem.time_setitem
```

### CODE REVIEW

**Comment by phofl** [comment]
cc @jbrockmendel would you mind having a look here?

**Comment by jbrockmendel** [comment]
will look over the weekend (applies to many many pings in my inbox)

File: pandas/core/internals/blocks.py [hunk 1]
```diff
@@ -1850,10 +1850,39 @@ def get_values(self, dtype: DtypeObj | None = None) -> np.ndarray:
     def values_for_json(self) -> np.ndarray:
         return self.values
 
-    def delete(self, loc) -> Block:
-        values = np.delete(self.values, loc, 0)
-        mgr_locs = self._mgr_locs.delete(loc)
-        return type(self)(values, placement=mgr_locs, ndim=self.ndim)
+    def delete(self, loc) -> list[Block]:
```
Comment (WillAyd) [inline pandas/core/internals/blocks.py:1853]: Feels a bit strange to have a Block method now return a list of itself - is it possible to maintain the same return type while still getting a performance boost?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/internals/blocks.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/internals/blocks.py:1853]: Lots of Block methods return `list[Block]`

Comment (WillAyd) [reply pandas/core/internals/blocks.py:1853]: Ah OK - sorry for the ignorance

File: pandas/core/internals/blocks.py [hunk 2]
```diff
@@ -1531,11 +1531,11 @@ def putmask(self, mask, new) -> list[Block]:
 
         return [self]
 
-    def delete(self, loc) -> Block:
+    def delete(self, loc) -> list[Block]:
```
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py:1534]: for 2D EAs we probably want to do the same thing as we do for NumpyBlock?

File: pandas/core/internals/blocks.py [hunk 3]
```diff
@@ -1850,10 +1850,39 @@ def get_values(self, dtype: DtypeObj | None = None) -> np.ndarray:
...
+        # all columns after our last indexer if any
+        loc = np.concatenate([loc, [self.values.shape[0]]])
+        mgr_locs_arr = self._mgr_locs.as_array
+        new_blocks: list[Block] = []
+
+        previous_loc = -1
+        for idx in loc:
+
+            if idx == previous_loc + 1:
+                # There is now column between current and last idx
```
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py:1875]: should "now" here be "no"?

File: pandas/core/internals/blocks.py [hunk 4]
```diff
@@ -1850,10 +1850,39 @@ def get_values(self, dtype: DtypeObj | None = None) -> np.ndarray:
...
+            raise IndexError
+
+        # Add one out-of-bounds indexer as maximum to collect
+        # all columns after our last indexer if any
+        loc = np.concatenate([loc, [self.values.shape[0]]])
+        mgr_locs_arr = self._mgr_locs.as_array
+        new_blocks: list[Block] = []
+
+        previous_loc = -1
+        for idx in loc:
```
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py:1872]: this reminds me of `BlockManager._slice_take_blocks_ax0` with `only_slice=True` if we passed e.g. `slice_or_indexer = np.delete(np.arange(length), loc))`.  is it worth trying to de-duplicate?

File: pandas/core/internals/blocks.py (hunk 1 above)
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py:1853]: can you add a docstring fleshing out what you're doing here

File: pandas/core/internals/blocks.py (hunk 2 above)
Comment (phofl) [reply pandas/core/internals/blocks.py:1534]: Do we have any 2D EAs? I agree in general, but thought all of them were 1D

File: pandas/core/internals/blocks.py (hunk 3 above)
Comment (phofl) [reply pandas/core/internals/blocks.py:1875]: Yes, thx

File: pandas/core/internals/managers.py [hunk 5]
```diff
@@ -1235,12 +1235,24 @@ def value_getitem(placement):
...
+                        + nbs_tup
                     )
                     self.blocks = blocks_tup
-                    self._blklocs[nb.mgr_locs.indexer] = np.arange(len(nb))
+                    self._blklocs[first_nb.mgr_locs.indexer] = np.arange(len(first_nb))
+
+                    for i, nb in enumerate(nbs_tup):
+                        self._blklocs[nb.mgr_locs.indexer] = np.arange(len(nb))
+                        self._blknos[nb.mgr_locs.indexer] = i + nr_blocks
                     # blk.delete gives a copy, so we can remove a possible reference
```
[the remark this answers is not in the document]
Comment (phofl) [reply pandas/core/internals/managers.py:1256]: I think we do return block copies? Was not sure if this was referring to the block or the underlying array

File: pandas/core/internals/blocks.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/internals/blocks.py:1534]: see NDArrayBackedExtensionBlock

File: pandas/core/internals/managers.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/internals/managers.py:1256]: im pretty sure the "reference" here is referring to arrays

File: pandas/core/internals/blocks.py (hunk 4 above)
Comment (phofl) [reply pandas/core/internals/blocks.py:1872]: Not sure, this seemed to do something different than what we need. Tried to get it to work a while back, but didn't get it to work

File: pandas/core/internals/managers.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (phofl) [reply pandas/core/internals/managers.py]: Got it, removed it

File: pandas/tests/frame/test_stack_unstack.py [hunk 6]
```diff
@@ -1323,8 +1323,23 @@ def test_unstack_non_slice_like_blocks(using_array_manager):
     # Case where the mgr_locs of a DataFrame's underlying blocks are not slice-like
 
     mi = MultiIndex.from_product([range(5), ["A", "B", "C"]])
-    df = DataFrame(np.random.randn(15, 4), index=mi)
-    df[1] = df[1].astype(np.int64)
+    df = DataFrame(
+        {
+            0: np.random.randn(
+                15,
```
[the remark this answers is not in the document]
Comment (phofl) [reply pandas/tests/frame/test_stack_unstack.py]: nope, this looks like a black bug. Fixed

File: pandas/core/internals/managers.py [hunk 7]
```diff
@@ -1235,14 +1235,24 @@ def value_getitem(placement):
                 if len(val_locs) == len(blk.mgr_locs):
                     removed_blknos.append(blkno_l)
                 else:
-                    nb = blk.delete(blk_locs)
+                    nbs = blk.delete(blk_locs)
+                    # Add first block where old block was and remaining blocks at
+                    # the end to avoid updating all block numbers
+                    first_nb = nbs[0]
+                    nbs_tup = tuple(nb for nb in nbs[1:])
```
Comment (mroeschke) [inline pandas/core/internals/managers.py:1242]: Nit: `tuple(nbs[1:])` to avoid iteration cost?

**Comment by jbrockmendel** [comment]
looks like some mypy complaints

File: pandas/core/internals/managers.py (hunk 7 above)
Comment (phofl) [reply pandas/core/internals/managers.py]: Thx, fixed

**Comment by phofl** [comment]
Thx, fixed mypy

**Review by jbrockmendel — APPROVED** [review]
LGTM cc @mroeschke

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @phofl

**Pull request merged**